File: solutions/221_maximal_square.py
'''
Given a 2D binary matrix filled with 0's and 1's, find the largest square containing only 1's and return its area.
Example:
Input: 
1 0 1 0 0
1 0 1 1 1
1 1 1 1 1
1 0 0 1 0
Output: 4

'''

class Solution:
    def maximalSquare(self, matrix):
        if 0 == len(matrix):
            return 0
        ans = 0
        for i in range(1, len(matrix)):
            if '1' == matrix[i][0]: ans = max(ans, 1)
            for j in range(1, len(matrix[0])):
                if '1' == matrix[0][j]: ans = max(ans, 1)
                if int(matrix[i][j]) > 0 and int(matrix[i - 1][j - 1]) > 0:
                    matrix[i][j] = min([
                        int(n) for n in (matrix[i - 1][j - 1],
                                         matrix[i][j - 1], matrix[i - 1][j])
                    ]) + 1
                    ans = max(ans, matrix[i][j])
        return ans * ans

matrix = [
    ['1', '0', '1', '0', '0'],
    ['1', '0', '1', '1', '1'],
    ['1', '1', '1', '1', '1'],
    ['1', '0', '0', '1', '0'],
]

print(Solution().maximalSquare(matrix))

File: _site/solutions/221_maximal_square.cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <map>
#include <stdio.h>
#include <vector>
/**
Given a 2D binary matrix filled with 0's and 1's, find the largest square
containing only 1's and return its area.
Example:
Input:
1 0 1 0 0
1 0 1 1 1
1 1 1 1 1
1 0 0 1 0
Output: 4

 **/
using namespace std;

static int speed_up = []() {
  std::ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();

class Solution {
public:
  int maximalSquare(vector<vector<char>> &matrix) {
    int rsize = matrix.size(), csize = matrix[0].size(), ans = 0;
    for (int i = 0; i < rsize; i++) {
      for (int j = 0; j < csize; j++) {
        ans = max(ans, squareSize(i, j, matrix));
      }
    }
    return ans * ans;
  }

  int squareSize(int a, int b, vector<vector<char>> &matrix) {
    if ('0' == matrix[a][b])
      return 0;
    int ans = 1;
    while (a + ans <= matrix.size() && b + ans <= matrix[0].size()) {
      for (int i = a; i < a + ans; i++) {
        for (int j = b; j < b + ans; j++)
          if ('0' == matrix[i][j])
            return ans - 1;
      }
      ans += 1;
    }
    return ans - 1;
  }
};
int main() {
  Solution s;
  vector<vector<char>> matrix = {
      {'1', '0', '1', '0', '0'},
      {'1', '0', '1', '1', '1'},
      {'1', '1', '1', '1', '1'},
      {'1', '0', '0', '1', '0'},
  };
  cout << s.maximalSquare(matrix) << endl;
}
